File: ex841.java
import java.util.List;

class Solution {
    private int[] visited;
    
    private void dfs(List<List<Integer>> rooms, int room) {
        if (visited[room] == 1)    return;
        visited[room] = 1;
        
        for (int key : rooms.get(room)) {
            dfs(rooms, key);
        }
    }
    
    public boolean canVisitAllRooms(List<List<Integer>> rooms) {
        visited = new int[rooms.size()];
        
        // Go dfs search
        dfs(rooms, 0);
        
        for (int i = 0; i < visited.length; i++)
            if (visited[i] == 0)    return false;
        
        return true;
    }
}

File: code/ex841.cpp
class Solution {
    int n;
    unordered_set<int> keys;
public:
    bool canVisitAllRooms(vector<vector<int>>& rooms) {
        n = rooms.size();    if (n <= 1)    return 1;
        
        vector<bool> visited(n);
        visited[0] = 1;
        keys.insert(0);
        return dfs(rooms, visited, 0);
    }
    
private:
    bool dfs(vector<vector<int>>& rooms, vector<bool>& visited, int cur) {
        for (int key : rooms[cur]) {
            keys.insert(key);
            if (keys.size() == n)    return 1;
            if (visited[key])    continue;
            visited[key] = 1;
            bool ret = dfs(rooms, visited, key);
            if (ret)    return 1;
        }
        return 0;
    }
};